### PR TITLE
Python 3.10 Union operator | support for JIT

### LINKED ISSUE #41412: [jit] Support for union types
[PEP 484](https://www.python.org/dev/peps/pep-0484/) specifies Unions as part of the type system. We should consider supporting them in the TorchScript type system, as they are a common pattern in Python code. This is a low-pri request from PyPER as well.

### LINKED ISSUE #101777: Python 3.10 Union operator `|` is not supported in jit script
### 🐛 Describe the bug

When using the new python 3.10 style of type annotations for unions (`|`) as defined in https://peps.python.org/pep-0604/ with `torch.jit.script`, a `RuntimeError` is raised.

For the given minimal example:
```python
from typing import Union
import torch

@torch.jit.script
def jit_test(
    ok: Union[int, float], bad: int | float
) -> None:
    pass
```

The traceback is:
```
(env) ➜  torch_test python test.py
Traceback (most recent call last):
  File "/Users/awb/torch_test/test.py", line 6, in <module>
    def jit_test(
  File "/Users/awb/torch_test/env/lib/python3.10/site-packages/torch/jit/_script.py", line 1341, in script
    fn = torch._C._jit_script_compile(
RuntimeError: 
Expression of type | cannot be used in a type expression:
  File "/Users/awb/torch_test/test.py", line 7
@torch.jit.script
def jit_test(
    ok: Union[int, float], bad: int | float
                                ~~~~~~~~~~~ <--- HERE
) -> None:
    pass
```

We would expect that the `bad` argument is treated identically as the `ok` argument since their typing is functionally equivalent.

### Versions

```
Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.11 (main, May 17 2023, 17:29:08) [Clang 14.0.3 (clang-1403.0.22.14.1)] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[conda] Could not collect
```

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### PR DESCRIPTION
**Pull request by randolf-scholz**
Fixes #101777

- [x] Duplicated the tests from `test/jit/test_union.py` into [`test/jit/test_union_pep604.py`](https://github.com/pytorch/pytorch/pull/109293/files#diff-b981f6493093482b43b0e62057b0c01b004b3e932d4e63a1166c3808c0172b83), using PEP604 style Unions
- [x] Exchanged custom `get_args` and `get_origin`  with `typing.get_args` and `typing.get_origin` which have the same functionality and became part of the standard library in 3.8
- [x] Added utility function `pep604union_to_union` in `tree_views.h` which converts a `BinOP("|")` node into the corresponding `Union`. This function intercepts `ScriptTypeParser::parseTypeFromExpr` and `ScriptTypeParser::parseTypeFromExprImpl` and patches the expression.
- [ ] There is a single failing test, I commented it out for the moment to see if CI complains about anything else. I tried several hours to figure out how to patch it, but I am not experienced with C++ development and debugging.

From what I could gather, the following fails:

```python
    def test_union_optional_of_union_return(self):
        @torch.jit.script
        def fn() -> None | str | int:
            y: Optional[int | str] = "foo"
            return y
```

In the section:

https://github.com/pytorch/pytorch/blob/75b954b7157ef26102796606944210b1ba7555a3/torch/csrc/jit/frontend/script_type_parser.cpp#L232-L243

When using regular `Union`, the `resolver` path is taken, whereas with the patch pep604 union, `resolveType` doesn't work.

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
okey dokey

**Comment by randolf-scholz** [comment]
@ezyang This shouldn't be merged yet, first the failure of the Optional simplification should be clarified.

**Comment by randolf-scholz** [comment]
Basically this one works:

```python
@torch.jit.script
def fn() -> Union[None, str, int]:
    y: Optional[Union[int, str]] = "foo"
    return y
```

But this one fails

```python
@torch.jit.script
def fn() -> None | str | int:
    y: Optional[int | str] = "foo"
    return y
```

with 

```
Return value was annotated as having type Union[int, NoneType, str]
but is actually of type Optional[Union[int, str]]:
```

**Comment by randolf-scholz** [comment]
@ezyang  I couldn't quite figure out how to solve it. Either we find a solution or one can decide to not care about this error. Also if you haven't it would be good to check the C++ code as I am a novice with respect to C++ development.

**Review by randolf-scholz — COMMENTED** [review]
What to do about the failing Optional simplification?

File: torch/csrc/jit/frontend/tree_views.h [hunk 1]
```diff
@@ -1208,6 +1208,61 @@ struct Delete : public Stmt {
...
+    return expr;
+
+  // In order to support unions with more than 2 operands ((x|y)|z), we need to
+  // recursively flatten the tree of | expressions.
+  auto members = get_pep604_union_members(expr);
+  auto synthesised_union = Subscript::create(
+      expr.range(),
+      Var::create(expr.range(), Ident::create(expr.range(), "Union")),
+      List<Expr>::create(expr.range(), members));
+  return std::move(synthesised_union);
```
Comment (ezyang) [inline torch/csrc/jit/frontend/tree_views.h:1263]: nit: just `return Subscript::create` directly. There's a long thing about return on std::move impeding copy elision but iirc Exprs are cheap to move so it doesn't really matter.

**Comment by ezyang** [comment]
I'm willing to accept this with bugs. It will just mean some valid programs fail to typecheck, right? You can always figure it out later and send us a patch.

We don't have dedicated TS cycles so I am mostly estimating that this feature is relatively low risk to add. I also am not sure how to solve your particular problem, but if this would be useful to you it's better to just get this in.

File: torch/csrc/jit/frontend/tree_views.h (hunk 1 above)
Comment (randolf-scholz) [reply torch/csrc/jit/frontend/tree_views.h:1263]: I added this after it was suggested in failed CI: https://github.com/pytorch/pytorch/actions/runs/6209430790/job/16857935849#step:10:13787

**Comment by randolf-scholz** [comment]
> I'm willing to accept this with bugs. It will just mean some valid programs fail to typecheck, right? You can always figure it out later and send us a patch.
> 
> We don't have dedicated TS cycles so I am mostly estimating that this feature is relatively low risk to add. I also am not sure how to solve your particular problem, but if this would be useful to you it's better to just get this in.

It means some valid programs fail JIT-compilation, specifically, this test fails (in the duplicated tests from `test/jit/test_union.py`):

```pyhton
    def test_union_optional_of_union_is_flattened(self):
        @torch.jit.script
        def fn(flag: int) -> str | int | None:
            y: int | str | None = "foo"
            if flag == 0:
                x: Optional[int | str] = y
            elif flag == 1:
                x: Optional[int | str] = 1
            else:
                x: Optional[int | str] = None
            return x
```

with 

```
Return value was annotated as having type Union[int, NoneType, str]
but is actually of type Optional[Union[int, str]]:
```

It succeeds if `|`-unions are replaced by `typing.Union`.

**Comment by ezyang** [comment]
Yeah, this is OK by me. Essentially, you are only strictly expanding the set of programs we accept, there are some programs we don't manage to get but they are easy to work around.

File: test/jit/test_union_pep604.py [hunk 2]
```diff
@@ -0,0 +1,1063 @@
...
+        @torch.jit.script
+        def fn(flag: int) -> str | int | None:
+            y: int | str | None = "foo"
+            if flag == 0:
+                x: Optional[int | str] = y
+            elif flag == 1:
+                x: Optional[int | str] = 1
+            else:
+                x: Optional[int | str] = None
+            return x
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/jit/test_union_pep604.py:374]: Since we're just going to accept as is, instead of skipping I recommend expectedFailure'ing this test

**Pull request closed without merging**